Jetcaster/core/src/main/java/com/example/jetcaster/core/data/domain/FilterableCategoriesUseCase.kt

```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright 2024 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.example.jetcaster.core.data.domain
+
+import com.example.jetcaster.core.data.database.model.Category
+import com.example.jetcaster.core.data.model.FilterableCategory
+import com.example.jetcaster.core.data.repository.CategoryStore
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.combine
+import kotlinx.coroutines.flow.onEach
+
+/**
+ * Use case for categories that can be used to filter podcasts.
+ */
+class FilterableCategoriesUseCase(
+    private val categoryStore: CategoryStore
+) {
+    operator fun invoke(
+        selectedCategoryStateFlow: MutableStateFlow<Category?>
```

Comment:
I'm not a fan of mutating the stateflow this way as a side effect of this use case.
If I understand this correctly, the `selectedCategoryStateFlow` might not be needed in this use case anyway. We could have `onEach{}` in the viewmodel that uses this use case and mutate that `selectedCategoryFlow` there.

Reply:
Agreed that it would be good to remove this side effect. Will update.

Reply:
I realised the problem might be more in depth with how you modelled the object. 

The `FilterableCategory` has now `isSelected` flag, which I think doesn't represent the real state as (if I read it right) only one category can be selected. This way we add complexity to all of the domain models, where in reality this is just a UI layer selection. 

I think things would be simplified if you passed to the composables `selectedCategory: Category?` as well and just had  `categories: List<Category>` instead of `filterableCategories`.

Forgive me if I don't see something. I see now basically what I suggest was sort of there before 🤔 😅 .

Reply:
Ok, now I understand your point that you wanted to have this functionality be testable and actually part of the domain state of your app.

I still think there could be improvement instead.

Instead of having `isSelected` in each category, I would suggest having something like 
```
data class FilterableModel(
    val categories: List<Category>,
    val selectedCategory: Category?
)
```

This way the use case might be still quite simple without the lambda and would still contain the logic to select the first category. Something like:

```
        categoryStore.categoriesSortedByPodcastCount()
            .map {
                FilterableModel(
                    categories = it,
                    // If we haven't got a selected category yet, select the first
                    selectedCategory = selectedCategory ?: it.firstOrNull()
                )
            }
```

The view state in the viewmodel would either be as before with the `DiscoverViewState`, or as is now, but you would add `selectedCategory:Category?` all the way down where it's needed.

The use case tests can be simplified this way as well without the side effects:

```
    @Test
    fun whenNoSelectedCategory_onEmptySelectedCategoryInvoked() = runTest {
        val filterableCategories = useCase(null).first()
        assertEquals(
            filterableCategories.categories[0],
            filterableCategories.selectedCategory
        )
    }

    @Test
    fun whenSelectedCategory_correctFilterableCategoryIsSelected() = runTest {
        val selectedCategory = testCategories[2]
        val filterableCategories = useCase(selectedCategory)

        assertEquals(filterableCategories.first().selectedCategory, selectedCategory)
    }
```


What do you think? 😊